simulator/tools/rofi_hal_sim.cpp

```diff
@@ -0,0 +1,243 @@
+#include "rofi_hal_sim.hpp"
+
+#include <cassert>
+
+namespace rofi
+{
+    namespace hal
+    {
+        RoFI::RoFI() : rdata( boost::make_shared< Data >( 3 ) ) {}
+
+        RoFI::Joint RoFI::getJoint( int index )
+        {
+            if ( index < 0 || static_cast< size_t >( index ) > rdata->joints.size() )
+                throw std::out_of_range( "Joint index is out of range" );
+            return rdata->joints[ index ]->getJoint();
+        }
+
+        RoFI::Joint::Joint( RoFI::Joint::Data & data ) : jdata( &data ) {}
+
+        RoFI::Data::Data( int jointNumber )
+        {
+            for ( int i = 0; i < jointNumber; i++ )
+            {
+                joints.push_back( std::make_unique< Joint::Data >( *this, i ) );
+            }
+
+            node = gazebo::transport::NodePtr( new gazebo::transport::Node() );
+            node->Init();
+
+            std::string moduleName = "universalModule";
+
+            sub = node->Subscribe( "~/" + moduleName + "/response", &RoFI::Data::onResponse, this );
+            pub = node->Advertise< rofi::messages::RofiCmd >( "~/" + moduleName + "/control" );
+
+            std::cerr << "Waiting for connection...\n";
+            pub->WaitForConnection();
+            std::cerr << "Connected\n";
+        }
+
+        void RoFI::Data::onResponse( const RoFI::Data::RofiRespPtr & resp )
+        {
+            std::cerr << "Got response\n";
+            messages::RofiCmd::Type resptype = resp->resptype();
+            switch ( resptype )
+            {
+                case messages::RofiCmd::JOINT_CMD:
+                {
+                    int index = resp->jointresp().joint();
+                    if ( index < 0 || static_cast< size_t >( index ) > joints.size() )
+                        throw std::out_of_range( "Joint index is out of range" );
+                    joints[ index ]->onResponse( resp );
+                    break;
+                }
+                default:
+                    // TODO
+                    break;
+            }
+        }
+
+        RoFI::Joint RoFI::Joint::Data::getJoint()
+        {
+            return Joint( *this );
+        }
+
+        messages::RofiCmd RoFI::Joint::Data::getCmdMsg( messages::JointCmd::Type type ) const
+        {
+            messages::RofiCmd rofiCmd;
+            rofiCmd.set_cmdtype( messages::RofiCmd::JOINT_CMD );
+            rofiCmd.mutable_jointcmd()->set_joint( jointNumber );
+            rofiCmd.mutable_jointcmd()->set_cmdtype( type );
+            return rofiCmd;
+        }
+
+        std::future< RoFI::Data::RofiRespPtr > RoFI::Joint::Data::registerPromise( messages::JointCmd::Type type )
+        {
+            std::lock_guard< std::mutex > lg( respMapMutex );
```

Comment:
I prefer consistent naming of lock guards

Reply:
Changed all to `lock`.